File: project/drivers/driver/pascal/tests/IntegrationTest.pas
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

program IntegrationTest;

{$mode delphi}
{$APPTYPE CONSOLE}

uses
  {$IFDEF UNIX}
  cthreads,
  {$ENDIF}
  SysUtils, Variants,
  ScratchBird.Client, ScratchBird.Sql, ScratchBird.Metadata, ScratchBird.Protocol, ScratchBird.Errors;

var
  Dsn: string;
  StreamSql: string;
  CancelSql: string;
  GeneratedKeySql: string;
  GeneratedKeyExpectedText: string;
  Client: TScratchBirdClient;

procedure Fail(const MessageText: string);
begin
  raise Exception.Create(MessageText);
end;

procedure AssertTrue(Value: Boolean; const MessageText: string);
begin
  if not Value then
    Fail(MessageText);
end;

procedure AssertEqualInt64(Expected, Actual: Int64; const MessageText: string);
begin
  if Expected <> Actual then
    Fail(MessageText + ': expected=' + IntToStr(Expected) + ' actual=' + IntToStr(Actual));
end;

procedure AssertEqualUInt64(Expected, Actual: UInt64; const MessageText: string);
begin
  if Expected <> Actual then
    Fail(MessageText + ': expected=' + IntToStr(Int64(Expected)) + ' actual=' + IntToStr(Int64(Actual)));
end;

function RequireVariantInt64(const Value: Variant; const MessageText: string): Int64;
begin
  if VarIsNull(Value) or VarIsEmpty(Value) then
    Fail(MessageText + ': value is null/empty');
  try
    Result := VarAsType(Value, varInt64);
  except
    on E: Exception do
      Fail(MessageText + ': expected integer-convertible value (' + E.Message + ')');
  end;
end;

procedure DrainStream(Stream: TScratchBirdResultStream; out RowCount: Integer);
var
  Row: TArray<Variant>;
begin
  RowCount := 0;
  while True do
  begin
    Row := Stream.ReadRow;
    if Row = nil then
      Break;
    Inc(RowCount);
  end;
end;

function TryReadMetadataTextValue(const Row: TMetadataRow; const CandidateFields: array of string; out TextValue: string): Boolean;
var
  I: Integer;
  FieldValue: Variant;
begin
  for I := Low(CandidateFields) to High(CandidateFields) do
  begin
    if not MetadataRowTryGetValue(Row, CandidateFields[I], FieldValue) then
      Continue;
    if VarIsNull(FieldValue) or VarIsEmpty(FieldValue) then
      Continue;
    TextValue := Trim(VarToStr(FieldValue));
    if TextValue <> '' then
      Exit(True);
  end;
  TextValue := '';
  Result := False;
end;

procedure AssertMetadataRestrictionRoundTrip(AClient: TScratchBirdClient; const CollectionName, RestrictionKey: string;
  const CandidateFields: array of string);
var
  BaseRows: TMetadataRows;
  FilteredRows: TMetadataRows;
  Restrictions: TMetadataRow;
  TargetValue: string;
  RowValue: string;
  I: Integer;
begin
  BaseRows := AClient.QueryMetadataRows(CollectionName);
  if Length(BaseRows) = 0 then
  begin
    Writeln('MetadataRestrictionTest: SKIPPED (' + CollectionName + ' has no rows)');
    Exit;
  end;

  if not TryReadMetadataTextValue(BaseRows[0], CandidateFields, TargetValue) then
  begin
    Writeln('MetadataRestrictionTest: SKIPPED (' + CollectionName + ' does not expose expected filter field)');
    Exit;
  end;

  SetLength(Restrictions, 1);
  Restrictions[0].Name := RestrictionKey;
  Restrictions[0].Value := TargetValue;
  FilteredRows := AClient.QueryMetadataRows(CollectionName, Restrictions);
  AssertTrue(Length(FilteredRows) > 0, 'restricted QueryMetadataRows(' + CollectionName + ') should return rows');

  for I := 0 to High(FilteredRows) do
  begin
    if not TryReadMetadataTextValue(FilteredRows[I], CandidateFields, RowValue) then
      Fail('restricted QueryMetadataRows(' + CollectionName + ') row does not expose expected field');
    AssertTrue(SameText(RowValue, TargetValue),
      'restricted QueryMetadataRows(' + CollectionName + ') should match requested ' + RestrictionKey);
  end;
end;

procedure RequireMetadataCollectionHasColumnsAndExecutes(AClient: TScratchBirdClient; const CollectionName: string);
var
  Stream: TScratchBirdResultStream;
  RowCount: Integer;
begin
  Stream := AClient.QueryMetadata(CollectionName);
  try
    DrainStream(Stream, RowCount);
    AssertTrue(Length(Stream.Columns) > 0, CollectionName + ' should expose at least one metadata column');
    AssertTrue(RowCount >= 0, CollectionName + ' row count should be non-negative');
  finally
    Stream.Free;
  end;
end;

procedure RequireMetadataWrapperHasColumnsAndExecutes(AClient: TScratchBirdClient; const WrapperName: string);
var
  Stream: TScratchBirdResultStream;
  RowCount: Integer;
begin
  if WrapperName = 'catalogs' then
    Stream := AClient.GetCatalogs
  else if WrapperName = 'schemas' then
    Stream := AClient.GetSchemas
  else if WrapperName = 'tables' then
    Stream := AClient.GetTables
  else if WrapperName = 'columns' then
    Stream := AClient.GetColumns
  else if WrapperName = 'indexes' then
    Stream := AClient.GetIndexes
  else if WrapperName = 'index_columns' then
    Stream := AClient.GetIndexColumns
  else if WrapperName = 'constraints' then
    Stream := AClient.GetConstraints
  else if WrapperName = 'primary_keys' then
    Stream := AClient.GetPrimaryKeys
  else if WrapperName = 'foreign_keys' then
    Stream := AClient.GetForeignKeys
  else if WrapperName = 'table_privileges' then
    Stream := AClient.GetTablePrivileges
  else if WrapperName = 'column_privileges' then
    Stream := AClient.GetColumnPrivileges
  else if WrapperName = 'procedures' then
    Stream := AClient.GetProcedures
  else if WrapperName = 'functions' then
    Stream := AClient.GetFunctions
  else if WrapperName = 'routines' then
    Stream := AClient.GetRoutines
  else if WrapperName = 'type_info' then
    Stream := AClient.GetTypeInfo
  else
    Fail('unsupported wrapper: ' + WrapperName);
  try
    DrainStream(Stream, RowCount);
    AssertTrue(Length(Stream.Columns) > 0, WrapperName + ' wrapper should expose at least one metadata column');
    AssertTrue(RowCount >= 0, WrapperName + ' wrapper row count should be non-negative');
  finally
    Stream.Free;
  end;
end;

procedure RequireOptionalMetadataCollectionExecutes(AClient: TScratchBirdClient; const CollectionName: string);
var
  Stream: TScratchBirdResultStream;
  RowCount: Integer;
begin
  Stream := AClient.QueryMetadata(CollectionName);
  try
    DrainStream(Stream, RowCount);
    AssertTrue(RowCount >= 0, CollectionName + ' row count should be non-negative');
  finally
    Stream.Free;
  end;
end;

procedure RequireOptionalMetadataWrapperExecutes(AClient: TScratchBirdClient; const WrapperName: string);
var
  Stream: TScratchBirdResultStream;
  RowCount: Integer;
begin
  if WrapperName = 'procedures' then
    Stream := AClient.GetProcedures
  else if WrapperName = 'functions' then
    Stream := AClient.GetFunctions
  else if WrapperName = 'routines' then
    Stream := AClient.GetRoutines
  else
    Fail('unsupported optional wrapper: ' + WrapperName);
  try
    DrainStream(Stream, RowCount);
    AssertTrue(RowCount >= 0, WrapperName + ' wrapper row count should be non-negative');
  finally
    Stream.Free;
  end;
end;

procedure TestQueryAndPrepareBind(AClient: TScratchBirdClient);
var
  Stream: TScratchBirdResultStream;
  Row: TArray<Variant>;
  Param: TScratchBirdParamInput;
begin
  Stream := AClient.ExecuteQuery('SELECT 1');
  try
    Row := Stream.ReadRow;
    AssertTrue(Length(Row) > 0, 'SELECT 1 should return one column');
    AssertEqualInt64(1, RequireVariantInt64(Row[0], 'SELECT 1 value'), 'SELECT 1 payload value');
  finally
    Stream.Free;
  end;

  Param.Value := 42;
  Param.Obj := nil;
  Stream := AClient.ExecuteQueryParams('SELECT ?::INTEGER', [Param]);
  try
    Row := Stream.ReadRow;
    AssertTrue(Length(Row) > 0, 'prepare/bind should return one column');
    AssertEqualInt64(42, RequireVariantInt64(Row[0], 'prepare/bind value'), 'prepare/bind payload value');
  finally
    Stream.Free;
  end;
end;

procedure TestTransactionLifecycle(AClient: TScratchBirdClient);
var
  Stream: TScratchBirdResultStream;
  Row: TArray<Variant>;
begin
  AClient.BeginTransaction;
  try
    AClient.Savepoint('sp_live_1');
    Stream := AClient.ExecuteQuery('SELECT 11');
    try
      Row := Stream.ReadRow;
      AssertTrue(Length(Row) > 0, 'transaction query should return row');
      AssertEqualInt64(11, RequireVariantInt64(Row[0], 'transaction query value'), 'transaction query payload');
    finally
      Stream.Free;
    end;
    AClient.ReleaseSavepoint('sp_live_1');

    AClient.Savepoint('sp_live_2');
    AClient.RollbackToSavepoint('sp_live_2');
    AClient.Commit;
  except
    AClient.Rollback;
    raise;
  end;

  AClient.BeginTransaction;
  AClient.Rollback;
end;

procedure TestLiveBatchAndMulti(AClient: TScratchBirdClient);
var
  Batch: TScratchBirdBatchResults;
  Rowsets: TScratchBirdRowsets;
begin
  Batch := AClient.ExecuteBatch(['SELECT 101', 'SELECT 202']);
  AssertTrue(Length(Batch) = 2, 'ExecuteBatch should return one summary per statement');

  Rowsets := AClient.QueryMulti(['SELECT 303', 'SELECT 404']);
  AssertTrue(Length(Rowsets) = 2, 'QueryMulti should return one rowset per statement');
  AssertTrue(Length(Rowsets[0].Rows) > 0, 'QueryMulti first rowset should include rows');
  AssertTrue(Length(Rowsets[1].Rows) > 0, 'QueryMulti second rowset should include rows');
  AssertEqualInt64(303, RequireVariantInt64(Rowsets[0].Rows[0][0], 'QueryMulti first row value'), 'QueryMulti first row payload');
  AssertEqualInt64(404, RequireVariantInt64(Rowsets[1].Rows[0][0], 'QueryMulti second row value'), 'QueryMulti second row payload');
end;

procedure TestLiveStreamControlPath(AClient: TScratchBirdClient; const CustomSql: string);
var
  Stream: TScratchBirdResultStream;
  QuerySql: string;
  RowCount: Integer;
begin
  QuerySql := Trim(CustomSql);
  if QuerySql = '' then
    QuerySql := 'SELECT id FROM basic_table ORDER BY id';

  Stream := AClient.ExecuteQuery(QuerySql);
  try
    AClient.StreamControl(STREAM_RESUME, 64, 1000);
    DrainStream(Stream, RowCount);
    AssertTrue(RowCount >= 1, 'stream-control query should return at least one row');
  finally
    Stream.Free;
  end;
end;

procedure TestMetadataFamiliesAndRestrictions(AClient: TScratchBirdClient);
var
  SchemaRows: TMetadataRows;
  FilteredRows: TMetadataRows;
  Restrictions: TMetadataRow;
  SchemaValue: Variant;
  SchemaName: string;
begin
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'catalogs');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'schemas');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'tables');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'columns');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'indexes');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'index_columns');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'constraints');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'primary_keys');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'foreign_keys');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'table_privileges');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'column_privileges');
  RequireOptionalMetadataCollectionExecutes(AClient, 'procedures');
  RequireOptionalMetadataCollectionExecutes(AClient, 'functions');
  RequireOptionalMetadataCollectionExecutes(AClient, 'routines');
  RequireMetadataCollectionHasColumnsAndExecutes(AClient, 'type_info');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'catalogs');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'schemas');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'tables');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'columns');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'indexes');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'index_columns');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'constraints');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'primary_keys');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'foreign_keys');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'table_privileges');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'column_privileges');
  RequireOptionalMetadataWrapperExecutes(AClient, 'procedures');
  RequireOptionalMetadataWrapperExecutes(AClient, 'functions');
  RequireOptionalMetadataWrapperExecutes(AClient, 'routines');
  RequireMetadataWrapperHasColumnsAndExecutes(AClient, 'type_info');

  SchemaRows := AClient.QueryMetadataRows('schemas');
  AssertTrue(Length(SchemaRows) > 0, 'QueryMetadataRows(schemas) should return at least one row');
  if MetadataRowTryGetValue(SchemaRows[0], 'schema_name', SchemaValue) or
     MetadataRowTryGetValue(SchemaRows[0], 'TABLE_SCHEM', SchemaValue) or
     MetadataRowTryGetValue(SchemaRows[0], 'table_schema', SchemaValue) then
  begin
    AssertTrue((not VarIsNull(SchemaValue)) and (not VarIsEmpty(SchemaValue)), 'schema value should not be null');
    SchemaName := VarToStr(SchemaValue);
  end
  else
    Fail('QueryMetadataRows(schemas) row does not expose schema name field');
  AssertTrue(SchemaName <> '', 'schema value should not be empty');

  SetLength(Restrictions, 1);
  Restrictions[0].Name := 'schema';
  Restrictions[0].Value := SchemaName;
  FilteredRows := AClient.QueryMetadataRows('schemas', Restrictions);
  AssertTrue(Length(FilteredRows) > 0, 'restricted QueryMetadataRows(schemas) should return rows');
  if MetadataRowTryGetValue(FilteredRows[0], 'schema_name', SchemaValue) or
     MetadataRowTryGetValue(FilteredRows[0], 'TABLE_SCHEM', SchemaValue) or
     MetadataRowTryGetValue(FilteredRows[0], 'table_schema', SchemaValue) then
  begin
    AssertTrue((not VarIsNull(SchemaValue)) and (not VarIsEmpty(SchemaValue)), 'restricted schema value should not be null');
    AssertTrue(VarToStr(SchemaValue) = SchemaName, 'restricted schema value should match requested schema');
  end
  else
    Fail('restricted QueryMetadataRows(schemas) row does not expose schema name field');

  AssertMetadataRestrictionRoundTrip(AClient, 'catalogs', 'catalog', ['catalog_name', 'TABLE_CATALOG', 'catalog']);
  AssertMetadataRestrictionRoundTrip(AClient, 'tables', 'table', ['table_name', 'TABLE_NAME', 'table']);
  AssertMetadataRestrictionRoundTrip(AClient, 'columns', 'column', ['column_name', 'COLUMN_NAME', 'column']);
  AssertMetadataRestrictionRoundTrip(AClient, 'indexes', 'index', ['index_name', 'INDEX_NAME', 'index']);
  AssertMetadataRestrictionRoundTrip(AClient, 'index_columns', 'index', ['index_name', 'INDEX_NAME', 'index']);
  AssertMetadataRestrictionRoundTrip(AClient, 'constraints', 'constraint', ['constraint_name', 'CONSTRAINT_NAME', 'constraint']);
  AssertMetadataRestrictionRoundTrip(AClient, 'primary_keys', 'constraint',
    ['constraint_name', 'CONSTRAINT_NAME', 'constraint']);
  AssertMetadataRestrictionRoundTrip(AClient, 'foreign_keys', 'constraint',
    ['constraint_name', 'CONSTRAINT_NAME', 'constraint']);
  AssertMetadataRestrictionRoundTrip(AClient, 'table_privileges', 'table', ['table_name', 'TABLE_NAME', 'table']);
  AssertMetadataRestrictionRoundTrip(AClient, 'column_privileges', 'column', ['column_name', 'COLUMN_NAME', 'column']);
  AssertMetadataRestrictionRoundTrip(AClient, 'procedures', 'procedure',
    ['procedure_name', 'PROCEDURE_NAME', 'routine_name']);
  AssertMetadataRestrictionRoundTrip(AClient, 'functions', 'function',
    ['function_name', 'FUNCTION_NAME', 'routine_name']);
  AssertMetadataRestrictionRoundTrip(AClient, 'routines', 'procedure',
    ['routine_name', 'ROUTINE_NAME', 'procedure_name', 'function_name']);
  AssertMetadataRestrictionRoundTrip(AClient, 'type_info', 'type',
    ['type_name', 'TYPE_NAME', 'data_type_name', 'DATA_TYPE_NAME']);
end;

procedure TestTypeCoverageFixture(AClient: TScratchBirdClient);
var
  Stream: TScratchBirdResultStream;
  Row: TArray<Variant>;
begin
  Stream := AClient.ExecuteQuery('SELECT * FROM type_coverage');
  try
    Row := Stream.ReadRow;
    AssertTrue(Length(Row) > 0, 'type_coverage fixture should return at least one row');
  finally
    Stream.Free;
  end;
end;

procedure TestOptionalCancelPath(AClient: TScratchBirdClient; const SqlText: string);
var
  Stream: TScratchBirdResultStream;
begin
  if SqlText = '' then
    Exit;
  Stream := AClient.ExecuteQuery(SqlText);
  AClient.Cancel;
  try
    Stream.ReadRow;
    raise Exception.Create('Cancel did not interrupt query');
  except
    on E: Exception do
      Writeln('CancelTest: OK');
  end;
end;

procedure TestGeneratedKeyPath(AClient: TScratchBirdClient; const SqlText, ExpectedText: string);
var
  Stream: TScratchBirdResultStream;
  RowCount: Integer;
  Expected: UInt64;
  EffectiveSql: string;
begin
  EffectiveSql := Trim(SqlText);
  if EffectiveSql = '' then
  begin
    Writeln('GeneratedKeyTest: SKIPPED (SCRATCHBIRD_PASCAL_GENERATED_KEY_SQL not set)');
    Exit;
  end;

  Stream := AClient.ExecuteQuery(EffectiveSql);
  try
    DrainStream(Stream, RowCount);
    if not Stream.HasLastInsertId then
      Fail('generated-key SQL did not expose last insert id');
    AssertTrue(Stream.LastInsertId > 0, 'generated-key SQL should expose non-zero last insert id');
    if Trim(ExpectedText) <> '' then
    begin
      if not TryStrToQWord(Trim(ExpectedText), Expected) then
        Fail('SCRATCHBIRD_PASCAL_GENERATED_KEY_EXPECTED must be an unsigned integer');
      AssertEqualUInt64(Expected, Stream.LastInsertId, 'generated-key SQL expected last insert id');
    end;
  finally
    Stream.Free;
  end;
end;

begin
  Dsn := GetEnvironmentVariable('SCRATCHBIRD_PASCAL_URL');
  if Dsn = '' then
  begin
    Writeln('IntegrationTest: SKIPPED (SCRATCHBIRD_PASCAL_URL not set)');
    Halt(0);
  end;
  Client := TScratchBirdClient.Create;
  try
    Client.Connect(Dsn);
    TestQueryAndPrepareBind(Client);
    TestTransactionLifecycle(Client);
    TestLiveBatchAndMulti(Client);
    TestMetadataFamiliesAndRestrictions(Client);
    TestTypeCoverageFixture(Client);
    GeneratedKeySql := GetEnvironmentVariable('SCRATCHBIRD_PASCAL_GENERATED_KEY_SQL');
    GeneratedKeyExpectedText := GetEnvironmentVariable('SCRATCHBIRD_PASCAL_GENERATED_KEY_EXPECTED');
    TestGeneratedKeyPath(Client, GeneratedKeySql, GeneratedKeyExpectedText);
    StreamSql := GetEnvironmentVariable('SCRATCHBIRD_PASCAL_STREAM_SQL');
    TestLiveStreamControlPath(Client, StreamSql);
    CancelSql := GetEnvironmentVariable('SCRATCHBIRD_PASCAL_CANCEL_SQL');
    TestOptionalCancelPath(Client, CancelSql);
    Writeln('IntegrationTest: OK');
  finally
    Client.Free;
  end;
end.

File: project/drivers/driver/rust/tests/integration_test.rs
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

use std::collections::HashMap;
use std::env;
use std::time::Duration;

use scratchbird::sql::Params;
use scratchbird::types::{Param, Value};
use scratchbird::{Client, Config, ErrorKind};
use tokio::time::timeout;

fn is_not_supported(err: &scratchbird::Error) -> bool {
    err.kind == ErrorKind::NotSupported || err.sqlstate.as_deref() == Some("0A000")
}

#[tokio::test]
async fn basic_query() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = client.query("SELECT 1").await.unwrap();
    client.close().await;
    assert!(!result.rows.is_empty());
}

#[tokio::test]
async fn prepare_bind_query() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = timeout(
        Duration::from_secs(5),
        client.query_params("SELECT ?::INTEGER", Params::from(vec![Param::Int32(42)])),
    )
    .await
    .expect("prepare_bind_query timed out")
    .expect("prepare_bind_query returned query error");
    client.close().await;
    assert!(!result.rows.is_empty());
    match result.rows[0][0] {
        Value::Int32(v) => assert_eq!(v, 42),
        Value::Int64(v) => assert_eq!(v, 42),
        _ => panic!("unexpected value type"),
    }
}

#[tokio::test]
async fn types_fixture_query() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = client
        .query("SELECT COUNT(*) FROM type_coverage")
        .await
        .unwrap();
    client.close().await;
    assert!(!result.rows.is_empty());
}

#[tokio::test]
async fn transaction_stays_active_across_connect_commit_and_rollback() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    client.savepoint("sp_rust_bootstrap").await.unwrap();
    client.release_savepoint("sp_rust_bootstrap").await.unwrap();
    client.begin(None).await.unwrap();
    client.commit(None).await.unwrap();
    client.savepoint("sp_rust_after_commit").await.unwrap();
    client.release_savepoint("sp_rust_after_commit").await.unwrap();
    client.begin(None).await.unwrap();
    client.rollback(None).await.unwrap();
    client.savepoint("sp_rust_after_rollback").await.unwrap();
    client.release_savepoint("sp_rust_after_rollback").await.unwrap();
    client.close().await;
}

#[tokio::test]
async fn nested_begin_is_rejected_while_active() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    client.begin(None).await.unwrap();
    let err = client.begin(None).await.unwrap_err();
    client.close().await;
    assert_eq!(err.kind, ErrorKind::Transaction);
    assert_eq!(err.sqlstate.as_deref(), Some("25000"));
}

#[tokio::test]
async fn post_rollback_query_observes_actual_result_after_fresh_reopen() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    client.begin(None).await.unwrap();
    client.rollback(None).await.unwrap();
    let result = client.query("SELECT 2").await.unwrap();
    client.close().await;
    assert!(!result.rows.is_empty());
    match result.rows[0][0] {
        Value::Int32(v) => assert_eq!(v, 2),
        Value::Int64(v) => assert_eq!(v, 2),
        _ => panic!("unexpected value type"),
    }
}

#[tokio::test]
async fn query_multi_returns_independent_result_sets() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = client
        .query_multi(
            "SELECT 1 AS first_value; SELECT 2 AS second_value",
            Params::from(()),
        )
        .await;
    let result = match result {
        Ok(value) => value,
        Err(err) if is_not_supported(&err) => return,
        Err(err) => panic!("query_multi failed: {}", err),
    };
    client.close().await;
    assert_eq!(result.len(), 2);
    match result[0].rows[0][0] {
        Value::Int32(v) => assert_eq!(v, 1),
        Value::Int64(v) => assert_eq!(v, 1),
        _ => panic!("unexpected first result value type"),
    }
    match result[1].rows[0][0] {
        Value::Int32(v) => assert_eq!(v, 2),
        Value::Int64(v) => assert_eq!(v, 2),
        _ => panic!("unexpected second result value type"),
    }
}

#[tokio::test]
async fn execute_batch_returns_summary() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let batch = client
        .execute_batch(
            "SELECT ?::INTEGER AS value",
            vec![
                Params::Positional(vec![Param::from(11_i32)]),
                Params::Positional(vec![Param::from(22_i32)]),
            ],
        )
        .await;
    let batch = match batch {
        Ok(value) => value,
        Err(err) if is_not_supported(&err) => return,
        Err(err) => panic!("execute_batch failed: {}", err),
    };
    client.close().await;
    assert_eq!(batch.items.len(), 2);
    assert_eq!(batch.items[0].index, 0);
    assert_eq!(batch.items[1].index, 1);
}

#[tokio::test]
async fn call_executes_callable_escape_syntax() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = client
        .call(
            "{ ? = call abs(?) }",
            Params::Positional(vec![Param::from(-3_i32)]),
        )
        .await;
    let result = match result {
        Ok(value) => value,
        Err(err) if is_not_supported(&err) => return,
        Err(err) => panic!("call failed: {}", err),
    };
    client.close().await;
    assert!(!result.rows.is_empty());
    match result.rows[0][0] {
        Value::Int32(v) => assert_eq!(v, 3),
        Value::Int64(v) => assert_eq!(v, 3),
        _ => panic!("unexpected callable return value type"),
    }
}

#[tokio::test]
async fn execute_with_generated_keys_returns_vec() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = client
        .execute_with_generated_keys("SELECT 1 AS one", Params::from(()))
        .await;
    let result = match result {
        Ok(value) => value,
        Err(err) if is_not_supported(&err) => return,
        Err(err) => panic!("execute_with_generated_keys failed: {}", err),
    };
    client.close().await;
    assert!(result.is_empty() || result.iter().all(|id| *id >= 0));
}

#[tokio::test]
async fn query_metadata_with_restrictions_filters_schema_rows() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();

    let all_schemas = client.query_metadata("schemas").await.unwrap();
    if all_schemas.rows.is_empty() {
        client.close().await;
        return;
    }

    let schema_col = all_schemas.columns.iter().position(|col| {
        col.name.eq_ignore_ascii_case("schema_name")
            || col.name.eq_ignore_ascii_case("TABLE_SCHEM")
            || col.name.eq_ignore_ascii_case("table_schema")
    });
    let Some(schema_col) = schema_col else {
        client.close().await;
        return;
    };

    let target_schema = match all_schemas.rows[0].get(schema_col) {
        Some(Value::String(value)) if !value.is_empty() => value.clone(),
        _ => {
            client.close().await;
            return;
        }
    };

    let restrictions = HashMap::from([("schema".to_string(), target_schema.clone())]);
    let filtered = client
        .query_metadata_with_restrictions("schemas", &restrictions)
        .await
        .unwrap();
    client.close().await;

    assert!(!filtered.rows.is_empty());
    for row in filtered.rows {
        match row.get(schema_col) {
            Some(Value::String(value)) => assert_eq!(value, &target_schema),
            _ => panic!("unexpected schema value type"),
        }
    }
}

#[tokio::test]
async fn cancel_query() {
    let dsn = match env::var("SCRATCHBIRD_RUST_URL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let cancel_sql = match env::var("SCRATCHBIRD_RUST_CANCEL_SQL") {
        Ok(val) => val,
        Err(_) => return,
    };
    let config = Config::from_dsn(&dsn).unwrap();
    let mut client = Client::new(config);
    client.connect().await.unwrap();
    let result = timeout(Duration::from_millis(200), client.query(&cancel_sql)).await;
    if result.is_err() {
        let _ = client.cancel().await;
    }
    client.close().await;
}
